File: System/LifeSystem/LifeForm.py
from .conf import MIN_RESISTANCE, MAX_RESISTANCE, MIN_SKILL, MAX_SKILL

class _Skill:
    def __set_name__(self, owner, name):
        self._name = name

    def __get__(self, instance, owner):
        return instance.__dict__[self._name]
    
    def __set__(self, instance, value: int):
        value = max(value, MIN_SKILL)
        value = min(value, MAX_SKILL)
        instance.__dict__[self._name] = value

class LifeForm:

    speed = _Skill()
    strength = _Skill()

    def __init__(self, name: str, health: int, resistance: float, speed: int, strength: int):
        self._name = name
        self._health = health
        self._max_health = health
        self._resistance = resistance
        self.speed = speed
        self.strength = strength
        self._dead = False
        self.__verify__()

    def __verify__(self):
        if self._health < 0:
            self._death()
        
        self._resistance = max(self._resistance, MIN_RESISTANCE)
        self._resistance = min(self._resistance, MAX_RESISTANCE)
    
    def _death(self):
        self._health = 0
        self._max_health = 0
        self._resistance = 0
        self.speed = 0
        self.strength = 0
        self._dead = True

    @property
    def name(self):
        return self._name
    
    @name.setter
    def name(self, name: str):
        self._name = name
    
    @property
    def health(self):
        return self._health
    
    @health.setter
    def health(self, newHealth: int):
        if newHealth < 0:
            self._death()
        elif newHealth > self._max_health:
            self.recoverAllHealth()
        else:
            self._health = newHealth
    
    @property
    def max_health(self):
        return self._max_health
    
    @property
    def resistance(self):
        return self._resistance
    
    @property
    def isDead(self):
        return self._dead
    
    def takeDamage(self, damage: int):
        self.health -= damage
    
    def recoverAllHealth(self):
        self.health = self.max_health
    
    def recoverHealth(self, health):
        self.health += health

File: System/LifeSystem/NewLifeSystem/LifeForm.cpp
#include <ctime>
#include <cstdlib>
#include <cmath>

#include "LifeForm.hh"

LifeForm::LifeForm(std::string n, unsigned int h, float r, unsigned int sp, unsigned int st) :
    name(n), health(h), max_health(h), resistance(r), speed(sp), strength(st)
{
    dead = false;
    __verify__();
}

LifeForm::~LifeForm()
{
}

void LifeForm::__verify__() {
    if (resistance < MIN_RESISTANCE) {
        resistance = MIN_RESISTANCE;
    } else if (resistance > MAX_RESISTANCE) {
        resistance = MAX_RESISTANCE;
    }

    if (speed > MAX_SKILL) {
        speed = MAX_SKILL;
    }
    if (strength > MAX_SKILL) {
        strength = MAX_SKILL;
    }
}

std::string LifeForm::getName() {
    return name;
}

unsigned int LifeForm::getHealth() {
    return health;
}

unsigned int LifeForm::getMaxHealth() {
    return max_health;
}

float LifeForm::getResistance() {
    return resistance;
}

unsigned int LifeForm::getSpeed() {
    return speed;
}

unsigned int LifeForm::getStrength() {
    return strength;
}

void LifeForm::death() {
    max_health = 0;
    resistance = 0;
    speed = 0;
    strength = 0;
    dead = true;
    // May want to create and return a Corpse Class
}

void LifeForm::takeDamage(unsigned int damage) {
    if (damage == 0) {
        return;
    }
    if (health <= damage) {
        health = 0;
        this->death();
    } else {
        health -= damage;
    }
}

unsigned int LifeForm::attack() {
 if (!dead) {
    srand(time(0));
    return (unsigned int)round(rand() / (float)RAND_MAX * speed + rand() / (float)RAND_MAX * strength);
 }
}

void LifeForm::recoverHealth(unsigned int recover) {
    if (max_health <= health + recover) {
        this->recoverAllHealth();
    } else {
        health += recover;
    }
}

void LifeForm::recoverAllHealth() {
    health = max_health;
}
